website/docs/cdktf/concepts/constructs.mdx

```diff
@@ -8,25 +8,315 @@ description: >-
+    this._nextStep.connect(step);
+    this.addAccessPermissions(this._nextStep, step);
+    this._nextStep = step;
+  }
+
+  public finalize() {
+    this._nextStep.finalize();
+  }
+}
+
+class MyStack extends TerraformStack {
+  constructor(scope: Construct, name: string) {
+    super(scope, name);
+    const pipeline = new DataPipeline(this, "my-pipeline");
+    pipeline.addStep(new TransformInputPipelineStep(this));
+    pipeline.addStep(new AggregateValuesPipelineStep(this));
+    pipeline.addStep(new PipelineStoreResultStep(this));
+    pipeline.finalize();
+  }
+}
+```
+
+### You can codify security constraints earlier in the process
+
+You can improve the feedback cycles about security concerns for developers from tools later in the workflow like code review or CI solutions like sentinel to the point where the cofiguration is authored. This can be done through type hints (disallowing options that are not save on the type system level) or through validation. Validations allow dynamic verifications and are shown to users as hints during `cdktf synth`, if they are on the error level they also fail the synth.
+
+#### Through Types
+
+```ts
+import { Construct } from "constructs";
+import { Ec2InstanceConfig as GeneratedEc2InstanceConfig } from "./.gen/providers/aws/lib/ec2-instance";
+
+enum AllowedInstanceSizes {
+  Nano = "t3.nano"
+  Micro = "t3.micro"
+  Small = "t3.small"
+  Medium = "t3.medium"
+}
+interface Ec2InstanceConfig extends GeneratedEc2InstanceConfig {
+  instanceSize: AllowedInstanceSizes;
+}
+
+class Ec2Instance extends Construct {
+  constructor(scope: Construct, name: string, config: Ec2InstanceConfig) {
+    super(scope, name, config);
+  }
+}
+```
+
+#### Through Validations
+
+```ts
+import { Construct } from "constructs";
+
+class S3Bucket extends Construct {
+  constructor(scope: Construct, name: string, config: S3BucketConfig) {
+    super(scope, name);
+
+    if (
+      this.isOpenToThePublic(config) &&
+      !config.tags["reason-why-this-is-public"]
+    ) {
+      Annotation.of(this).addWarning(
+        `This S3Bucket is open to the public and no reason why is stated in the tags, please add a "reason-why-this-is-public" tag`
+      );
+    }
+  }
+}
+```
+
+### Constructs and Testing
+
+Since Constructs can contain logic and are not just a data structure, they can be tested. This allows you to test the composition of your constructs and the logic they contain. This can be done with snapshot testing or unit testing.
+
+```ts
+import { Testing } from "cdktf";
+import MyApplicationsAbstraction from "../main";
+import { Ec2 } from "";
+
+
+describe("Construct is of correct compostion and plans successfully", () => {
+    // make assertions about the composition of constructs
+    it("has EC2 of correct size", () => {
+        expect(
+            Testing.synthScope((scope) => {
+                new MyApplicationsAbstraction(scope, "app", {});
+            }).toHaveResourceWithProperties(Ec2, {
```

Comment:
the matcher seems to be called on synthScope() and not on expect()?

Reply:
Oh yeah, I'll change that